src/org/rascalmpl/library/util/ErrorRecovery.rsc

```diff
@@ -0,0 +1,53 @@
+/**
+ * Copyright (c) 2024, NWO-I Centrum Wiskunde & Informatica (CWI)
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
+ *
+ * 2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
+ *
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **/
+ 
+module util::ErrorRecovery
+
+import ParseTree;
+import String;
+
+@synopsis{Check if a parse tree contains any error nodes, the result of error recovery.}
+bool hasErrors(Tree tree) = /appl(error(_, _, _), _) := tree;
```

Comment:
The language server for Pico uses this function [like this](https://github.com/usethesource/rascal-language-servers/blob/5a36693e1f3e1ada0e7cb1d54ce7849cbab074a3/rascal-lsp/src/main/rascal/demo/lang/pico/LanguageServer.rsc#L97-L98):

```rascal
// definitions of variables
rel[str, loc] defs = {<"<var.id>", var.src> | /IdType var := input, !hasErrors(var)};
```

Without context, it could be unclear that it's about *parse* errors. (This code in the Pico language server occurs in the definition of the analyzer, so it could also be about type errors.)

So, minor suggestion (primarily because it's user-facing): Maybe the function could be renamed to `hasParseErrors`. (And by the same token, maybe the module could be renamed to `ParseErrorRecovery`.)

Reply:
I'm looking for a solution where this function is not necessary. Since parse trees are a built-in feature of Rascal, there is opportunity for built-in features for error trees.

Reply:
There's a branch off of this one, with an experimental semantics for field projection on error trees. Pattern matching already fails naturally on them but field projection fails too often on error trees, making existing code less robust.